sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+      windowed_element = WindowedValue(element, timestamp, [window])
+    else:
+      element_and_restriction = (
+          value.value if isinstance(value, WindowedValue) else value)
+      assert isinstance(element_and_restriction, ElementAndRestriction)
+      element = element_and_restriction.element
+      restriction = element_and_restriction.restriction
+
+      if isinstance(value, WindowedValue):
+        windowed_element = WindowedValue(
+            element, value.timestamp, value.windows)
+      else:
+        windowed_element = WindowedValue(element, timestamp, [window])
+
+    tracker = self.sdf_invoker.invoke_create_tracker(restriction)
+    assert self._process_element_invoker
+    assert isinstance(self._process_element_invoker,
+                      SDFProcessElementInvoker)
+
+    output_values = self._process_element_invoker.invoke_process_element(
+        self.sdf_invoker, windowed_element, tracker)
+
+    sdf_result = None
+    for output in output_values:
+      if isinstance(output, SDFProcessElementInvoker.Result):
+        # SDFProcessElementInvoker.Result should be the last item yielded.
+        sdf_result = output
+        break
+      yield output
+
+    if not sdf_result.residual_restriction:
+      # All work for current residual and restriction pair is complete.
+      state.clear_state(window, self._element_tag)
+      state.clear_state(window, self._restriction_tag)
+      # Releasing output watermark by setting it to positive infinity.
+      state.add_state(window, self.watermark_hold_tag,
+                      WatermarkManager.WATERMARK_POS_INF)
+    else:
+      state.add_state(window, self._element_tag, element)
+      state.add_state(window, self._restriction_tag,
+                      sdf_result.residual_restriction)
+      state.add_state(window, self.watermark_hold_tag,
+                      WatermarkManager.WATERMARK_NEG_INF)
+      # Holding output watermark by setting it to negative infinity.
+      state.set_timer(
+          window, '', TimeDomain.WATERMARK, WatermarkManager.WATERMARK_NEG_INF)
+
+
+class SDFProcessElementInvoker(object):
+  """A utility that requsts checkpoints.
+
+  Requests a checkpoint after processing a predefined number of elements or
+  after a predefined time is elapsed from the process() invocation.
+  """
+
+  class Result(object):
+    def __init__(
+        self, residual_restriction=None, continuation=None,
+        future_output_watermark=None):
+      self.residual_restriction = residual_restriction
+      self.continuation = continuation
+      self.future_output_watermark = future_output_watermark
+
+  def __init__(
+      self, max_num_outputs, max_duration):
+    self._max_num_outputs = max_num_outputs
+    self._max_duration = max_duration
+    self._checkpoint_lock = Lock()
+
+  def test_method(self):
+    raise ValueError
+
+  def invoke_process_element(self, sdf_invoker, element, tracker):
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
Please add a docstring detailing the protocol here, particularly regarding the behavior of yielding the residual restriction as the last element.